config.default.yaml

```diff
@@ -106,6 +106,7 @@ cluster_options:
 
 build_shape_options:
   gadm_layer_id: 1 # GADM level area used for the gadm_shapes. Codes are country-dependent but roughly: 0: country, 1: region/county-like, 2: municipality-like
+  simplify_gadm: false # When true, shape polygons are simplified else no
```

Comment:
@davide-f Yes, I am planning to add it and it is already in progress. I was trying to understand the reason for the explosion of the regions_onshore.geojson file without simplification in the meantime.

Reply:
@SermishaNarayana @davide-f thanks a lot for taking care about that!

I'm afraid that it may not work with having a numerical option to simplify due to some `_simplify_polys( )` function itself... 

@jome1 did [a great investigation](https://github.com/pypsa-meets-earth/pypsa-earth/issues/1051#issue-2383439654) on behaviour of `_simplify_polys( )` which has demonstrated that all the polygons are being simplified independently of each other. That can lead to emerging a number of "enclaves" across the border of the regions.

I suspect explosion of the polygons you observed can be related to that: once we call `_simplify_polys( )` it results in emerging of large amount of such enclaves because the geometry is quite complex. The good news is that the issue is likely to be resolved with the next release of `shapely` which should contain an improved simplification algorithm

So, I'd probably leave a boolean flag for now and returned to the idea to add a numeric parameter, once an advanced simplification option will be available in `shapely`. What do you think?

Reply:
@SermishaNarayana you are absolutely right, that is `_simplify_polys( )` which creates issues in the #1051

Have misunderstood that you are observing the file expansion in case `_simplify_polys( )` is bypassed. Not sure if that's really related to #1051, then. Thank you so much for the explanation!

Just to be sure that I get the problem: you are also observing some strange geometry effects, right? If that is the case could you please post a picture there? Would be great to understand what is going on with geometries

Reply:
@ekatef Thanks for clarifying :)

The geometric effects that I observed are very similar to what has been mentioned as the issue with #1051 
 _simplify_polys()

Reply:
The other issue that I observed is in the clustered shape file (regions_onshore_elec_s_{clusters}.geojson)

In the runs for US, I have used 50 clusters (to be equal to number of US states). The observation is that there is an overlap in the statewise shapes itself. Some of the states are extending much beyond their actual borders

![Screenshot 2024-10-15 at 1 56 03 PM](https://github.com/user-attachments/assets/ce5cbc90-de57-4886-80d4-88af3d6ba832)

In the picture, the geometry of the region "US.26_1_AC" which refers to the state of Florida as tagged in the regions_onshore_elec_s_50.geojson file has been plotted. But the geometry of the state is projected to cover many more states of the US

Reply:
I am not sure but it might be indeed related to https://github.com/pypsa-meets-earth/pypsa-earth/issues/1051. I explain it myself, that due to the holes and overlaps in the fundamental/basis shapes, the clustering has some erros. The clustering happens on the basis of the shapes of the US states in this case, so the shapes of the US states itself should be flawless to receive flawless clusters.

Reply:
Thanks everyone.
Indeed this is weird and the combination of multiple issues:
1. the tolerance used for the gadm (when 0 or negative or False, no simplification is applied)
2. the algorithm in simplify polys that sometimes may lead to "normal" overlaps though not wanted
3. the same algorithm that sometimes lead to weird clustered shapes

In this PR we probably shall focus on 1 while tracking the other issues in the other PRs, otherwise it feels a bit overwhelming and hopefully along the way shapely improves the algorithm too.

I'd propose to use a numeric (or false) value instead of a pure boolean as it is an inexpensive feature that only improves the existing also giving quite flexibility.
With the boolean we only allow (a) enable the feature with predefined tolerance or (b) disabling the feature entirely that is a bit extreme.
With the number, the tolerance itself can be calibrated.

The tolerance has also implications into the geospatial comparison of subsequent rules, therefore the higher the simplification the better the performances, at the cost of accuracy though.
That's why customizing the number may be handy

Reply:
> In the picture, the geometry of the region "US.26_1_AC" which refers to the state of Florida as tagged in the regions_onshore_elec_s_50.geojson file has been plotted. But the geometry of the state is projected to cover many more states of the US

@SermishaNarayana could you share the geojson?
Do I understand well that the gadm file of florida contains the broader US? What is the large black shape you are showing? I understand the borders but not much the large states though

The original build_shapes should not overlap theoretically, but in practise it may be.

having slight overlap across boundaries may not lead to issues; alteratively, we can ensure no overlap in build_shapes with a heuristic, but not sure that is needed.

Reply:
@davide-f 

I have attached the regions_onshore_elec_s_50.geojson file when simplifying the GADM shapes with a tolerance of 0.01 (default)
[regions_onshore_elec_s_50_simplify_gadm.geojson.zip](https://github.com/user-attachments/files/17391828/regions_onshore_elec_s_50_simplify_gadm.geojson.zip)

The big black shape in the previous comment shows the geometric shape of the state of Florida as classified in this particular geojson file

Reply:
> @SermishaNarayana can you share also the gadm_shapes and the whole shapes folder? simplify_polys apply on gadm_shapes first and with alternative_clustering I expected that to be used.
The output of bus_regions is regions_onshore; then other rules apply and they edit them further.
You may have found another bug later on down the chain, but it is very likely that it is not linked to this PR

Reply:
@davide-f 

Sure, I have attached all the files in the shape folder
[shape_files.zip](https://github.com/user-attachments/files/17394720/shape_files.zip)

Also, with respect to using a numerical option for tolerance, I am planning to raise it as a separate PR. Is that alright? And also to clarify (a tolerance value of 0 has the same issues as not allowing the GADM shapes to be simplified )

Reply:
The gadm_shapes here are fine, so this PR should be ok. The problem is likely later on.

Let's also keep the flag true/false and open an issue for custom tolerances and minarea.

Feel free to turn this PR as ready for review whenever ready. Please add a line in release note :)

Reply:
@davide-f Apologies, The files I had uploaded earlier were the ones with simplify_gadm option turned ON. I am attaching another set of files with simplify_gadm options turned OFF.

[shape_files_non_simplified.zip](https://github.com/user-attachments/files/17395648/shape_files_non_simplified.zip)

Can you please verify if these files look alright? 
I shall add the release note :)

Reply:
Great! At first glance it looks good .
Here we have a great comparison and justification on why to add custom tolerances.
The feedback I see is the following:
1. the number of shapes in the two dataset is the same: 55 shapes in both. This means that the shapes are conserved.
2. the "simplified" one is 0.7Mb, while the non-simplified is 73 Mb. This is a good reason on why to use simplification. For some applications, however, the default value may be too much and lower simplication may be accepted.

The shapes are consistent and the bug you found later on in the simplification is not visible here. This suggests that the PR is ready to go functionality-wise :)